[RNMobile] Set and reference heading anchors

dcalhoun opened the pull request:
## Description
Address wordpress-mobile/gutenberg-mobile#1816 by adding the ability to set and reference Heading block anchors

- **Guntenberg Mobile:** wordpress-mobile/gutenberg-mobile#2947
- ~**WordPress iOS:** wordpress-mobile/WordPress-iOS#15567~
- ~**WordPress Android:** wordpress-mobile/WordPress-Android#13705~

## How has this been tested?
1. Edit a post and add Heading block.
1. Tap block settings button (i.e. settings cog icon).
1. Set value for "HTML Anchor" control.
1. Dismiss block settings.
1. Add Paragraph block with text including a link to the HTML Anchor from
   earlier step.
1. Publish post.
1. Visit published post in a web browser.
1. Tap link.

**Outcome:** Browser should scroll to Heading anchor.

## Screenshots

<details>
<summary>UI Examples</summary>

| iOS | Android |
| --- | --- |
| <img width="584" alt="ios-heading-selected" src="https://user-images.githubusercontent.com/438664/103777276-41d59e80-4ff6-11eb-9a37-905020c7333f.png"> | <img width="602" alt="android-heading-selected" src="https://user-images.githubusercontent.com/438664/103777296-48641600-4ff6-11eb-891a-cf443ab8b960.png"> |
| <img width="495" alt="ios-without-heading-anchor" src="https://user-images.githubusercontent.com/438664/104748406-53febd80-5717-11eb-8a24-f861da70b074.png"> | <img width="602" alt="android-without-heading-anchor" src="https://user-images.githubusercontent.com/438664/104748420-595c0800-5717-11eb-9b82-8242a52e360e.png"> |
| <img width="495" alt="ios-with-heading-anchor" src="https://user-images.githubusercontent.com/438664/104748468-65e06080-5717-11eb-898b-31bb59b2daf3.png"> | <img width="602" alt="android-with-heading-anchor" src="https://user-images.githubusercontent.com/438664/104748483-6b3dab00-5717-11eb-9ea3-551c9ec99d4a.png"> |

</details>

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

dcalhoun committed:
Allow setting heading anchor

dcalhoun committed:
Update change log

dcalhoun committed:
Disable auto capitalization for anchor text input

dcalhoun committed:
Update title and placeholder text

Address design feedback.

hypest commented:
👋 David! May I suggest changing the "prefix" in the PR title to `[RNMobile]` instead of just [Mobile]? "RNMobile" is not a hugely better prefix but, it's rather unique at that and disambiguating between mobile-web and native-mobile. Thanks!

dcalhoun commented:
@hypest I do not have permissions to manage the reviewers list. After consideration, including four members is likely unnecessary. Would please update the reviewers to list to be @ceyhun and @talldan please?

@ceyhun would you please review that my implementation works as expected and is a sound approach?

@talldan would you be willing to review my work to ensure my implementation does not disrupt any web experiences or violate any general project coding practices?

talldan commented:
@dcalhoun Thanks for the ping.

I tested this and on the web I'm seeing two way to set an anchor on the heading block now:
<img width="284" alt="Screenshot 2021-01-14 at 10 25 38 am" src="https://user-images.githubusercontent.com/677833/104536389-fe60dd00-5652-11eb-8e5a-df07c031c366.png">

The UI for the anchor in the advanced section is added to some blocks by the `supports` setting:
https://github.com/WordPress/gutenberg/blob/e1c3dcf26506d5b9fab2ed38da60c2fca471b6ca/packages/block-library/src/heading/block.json#L25

And the code that picks up that setting and adds the UI is here (added using filters):
https://github.com/WordPress/gutenberg/blob/e1c3dcf26506d5b9fab2ed38da60c2fca471b6ca/packages/block-editor/src/hooks/anchor.js#L64-L113

I think if you're looking to add anchor support on mobile, you'd probably want to use the same system to ensure it automatically works across all blocks that support the anchor setting.

ceyhun reviewed (commented):
Just saw the comment from Daniel. I guess the mobile implementation will be different from the web one, so I think it makes sense to review and test after that change. Feel free to ping me again 🙇

hypest commented:
> I tested this and on the web I'm seeing two way to set an anchor on the heading block now:

Aha, nice find there Daniel!

> I think if you're looking to add anchor support on mobile, you'd probably want to use the same system to ensure it automatically works across all blocks that support the anchor setting.

👋 @dcalhoun , just wanted to add that in terms of scope, it'd be cool if we can limit to the heading block for now, but still have an implementation that lends itself to expanding to other blocks. That said, I'm not sure if that'd be more practical or quicker than just go for all blocks as Daniel mentions, but if we can break up the scope in such a way that we land things in separate PRs it can help with moving forward in a nice pace. Thanks!

dcalhoun commented:
> > I tested this and on the web I'm seeing two way to set an anchor on the heading block now:
> 
> Aha, nice find there Daniel!

Indeed. Thank you for reviewing, Daniel. You have opened my eyes to a much wider view of the code base and it's incredibly helpful.

> it'd be cool if we can limit to the heading block for now, but still have an implementation that lends itself to expanding to other blocks. That said, I'm not sure if that'd be more practical or quicker than just go for all blocks as Daniel mentions, but if we can break up the scope in such a way that we land things in separate PRs it can help with moving forward in a nice pace.

@hypest whether it is more or less work to scope enabling anchors for headings will likely depend upon the response to my [questions for Thomas](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1816#issuecomment-760539916). If we is OK with using a generic heading for the bottom sheet, then enabling anchor hook for all blocks will likely "just work out of the box."

That said, I can go ahead and execute enabling anchors for Heading blocks only to sustain project momentum.

dcalhoun committed:
Fix broken documentation link

dcalhoun committed:
Refactor heading anchors for mobile

Previous approach ignored web implementation and resulted in double
anchor controls for web editor. This leverages the existing anchor hook
to fix the issue and share more code.

dcalhoun committed:
Build latest documentation

dcalhoun commented:
@talldan @ceyhun I have addressed the initial feedback. I believe this PR is now ready for review again. Thank you!

talldan commented on packages/block-editor/src/hooks/anchor.js:
```diff
@@ -67,41 +68,56 @@ export const withInspectorControl = createHigherOrderComponent(
+						) }
+						{ ! isWeb && props.name === 'core/heading' && (
+							<InspectorControls>
+								<PanelBody title={ __( 'Heading settings' ) }>
```
'Heading settings' might not be very scalable for the future, as other features that get added to the block inspector for Heading can't be put inside this panel. There's no way to access it from the Block's edit function.

dcalhoun replied:
Agreed, we'll need a more scalable heading in the future. [Thomas requested](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1816#issuecomment-756302170) that we use the "Heading settings" title for the time being.

hypest replied:
> Thomas requested

[My comment below](https://github.com/WordPress/gutenberg/pull/27935#discussion_r561787532) is actually covering the process here too, thanks!

talldan commented on packages/block-editor/src/hooks/anchor.js:
```diff
@@ -67,41 +68,56 @@ export const withInspectorControl = createHigherOrderComponent(
+								{ textControl }
+							</InspectorAdvancedControls>
+						) }
+						{ ! isWeb && props.name === 'core/heading' && (
```
Is there a plan to remove the restriction to just the one block? 

I'd hope that as long as `TextControl` works well on mobile, it should be fairly straightforward to support this on all blocks that declare support for an HTML Anchor.

dcalhoun replied:
Yes, I believe the plan is to implement other blocks in the future. For the time being, [Thomas and Stefanos requested](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1816#issuecomment-763199224) that we scope the work to Heading blocks.

From my testing, you are correct that wider anchor support is straightforward.

hypest replied:
> requested

Just a quick note on the process, as "requested" doesn't represent it very well: The way we work, at least on the native mobile apps, we make sure people collaborate by sharing information and context to design/implement solutions, including carving out the plan to get there.

We try to keep PR scope to the minimum so we can move land them in small but predictable steps (even if, say, landing means merging to a feature branch). We re-evaluate our plans as we go, when new info arises, like for example when we understand scope is actually too big for a single PR.

In our case here with the anchor feature, limiting to just the "Heading" block helps us pin down the implementation on mobile, add value by shipping the feature to the users, all while we making plans for the next steps.

Hope this helps, thanks!

dcalhoun replied:
@hypest thank you for the feedback and context.

ceyhun replied:
Maybe we can add a comment here linking the issue where we plan to implement `InspectorAdvancedControls` and/or enable anchors for all the other blocks?

talldan commented on packages/block-editor/README.md:
```diff
@@ -394,7 +394,7 @@ Undocumented declaration.
 _Related_
 
--   <https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/inspector-controls-advanced/README.md>
+-   <https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/inspector-advanced-controls/README.md>
```
good catch!

talldan reviewed (commented):
Things look good from a web point of view.  Happy for someone more familiar with mobile to do a test there and give the final thumbs up 👍.

dcalhoun commented on packages/block-editor/src/hooks/anchor.js:
```diff
@@ -67,41 +68,56 @@ export const withInspectorControl = createHigherOrderComponent(
+							</InspectorAdvancedControls>
+						) }
+						{ ! isWeb && props.name === 'core/heading' && (
+							<InspectorControls>
```
I agree. Ideally we reuse `InspectorAdvancedControls` for mobile once we have a design for display its entirety. Thomas and I had [a discussion](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1816#issuecomment-752257427) about the "Advanced" section, but opted to target only the anchor input for the initial implementation.

dcalhoun commented:
@talldan thank you again for the solid review. You shared great thoughts and questions. I have created #28363 to allow us to continue a discussion on when/how we might add wider support anchors on mobile.

ceyhun commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ For each user feature we should also add a importance categorization label  to i
 
 * [***] Add support for cross-posting between sites
 * [***] Full-width and wide alignment support for Columns
+* [**] Add support for setting heading anchors
```
We may need to move this note under `Unreleased` at this point.

dcalhoun replied:
Good catch! Done.

ceyhun reviewed (approved):
Sorry for the late review @dcalhoun, it LGTM 👍 Tested on iOS and Android and everything is working as expected. Thanks!

dcalhoun committed:
Improve release notes and documentation

Address code review feedback.

dcalhoun commented:
@ceyhun thank you for the review. I pushed a new commit to address your feedback. I believe this PR is ready to merge. I do not have the privileges required to rerun the failing CI tests or merge the PR. Would you be willing to help me do that please?

ceyhun commented:
I restarted the `End-to-End Tests / Admin` jobs, but they are already failing on master, so I think it'll be OK to merge this PR even if they don't pass.
